src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1259,52 +1284,107 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
 
             Debug.Assert((_readPos == 0 && _readLen == 0 && _writePos >= 0) || (_writePos == 0 && _readPos <= _readLen), "We're either reading or writing, but not both.");
+            Debug.Assert(!_isPipe || (_readPos == 0 && _readLen == 0), "Win32FileStream must not have buffered data here!  Pipes should be unidirectional.");
 
-            if (_isPipe)
+            bool writeDataStoredInBuffer = false;
+            if (!_isPipe) // avoid async buffering with pipes, as doing so can lead to deadlocks (see comments in ReadInternalAsyncCore)
             {
-                // Pipes are tricky, at least when you have 2 different pipes
-                // that you want to use simultaneously.  When redirecting stdout
-                // & stderr with the Process class, it's easy to deadlock your
-                // parent & child processes when doing writes 4K at a time.  The
-                // OS appears to use a 4K buffer internally.  If you write to a
-                // pipe that is full, you will block until someone read from 
-                // that pipe.  If you try reading from an empty pipe and 
-                // Win32FileStream's ReadAsync blocks waiting for data to fill it's 
-                // internal buffer, you will be blocked.  In a case where a child
-                // process writes to stdout & stderr while a parent process tries
-                // reading from both, you can easily get into a deadlock here.
-                // To avoid this deadlock, don't buffer when doing async IO on
-                // pipes.   
-                Debug.Assert(_readPos == 0 && _readLen == 0, "Win32FileStream must not have buffered data here!  Pipes should be unidirectional.");
+                // Ensure the buffer is clear for writing
+                if (_writePos == 0)
+                {
+                    if (_readPos < _readLen)
+                    {
+                        FlushRead();
+                    }
+                    _readPos = 0;
+                    _readLen = 0;
+                }
 
-                if (_writePos > 0)
-                    FlushWrite(false);
+                // Determine how much space remains in the buffer
+                int remainingBuffer = _bufferSize - _writePos;
+                Debug.Assert(remainingBuffer >= 0);
 
-                return WriteInternalCoreAsync(array, offset, numBytes, cancellationToken);
-            }
+                // Simple/common case:
+                // - The write is smaller than our buffer, such that it's worth considering buffering it.t
+                // - There's no active flush operation, such that we don't have to worry about the existing buffer being in use.
+                // - And the data we're trying to write fits in the buffer, meaning it wasn't already filled by previous writes.
+                // In that case, just store it in the buffer.
+                if (numBytes < _bufferSize && !HasActiveBufferOperation && numBytes <= remainingBuffer)
```

Comment:
Correct.  If you're only ever writing out >= to the buffer size, we can avoid allocating the buffer at all.  And even if the buffer needs to be allocated, by skipping the buffer for the == case, we can kick off the async operation now rather than deferring it until later, which is good since we know we want to do this and we may as well overlap it as much as possible with whatever we're going to do next.  The synchronous Write already skips the buffer for ==; this brings WriteAsync in line with it.